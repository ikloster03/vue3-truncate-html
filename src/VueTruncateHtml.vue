<!-- eslint-disable vue/no-v-html -->
<template>
  <div :class="proxyClasses.container">
    <div
      v-if="isHTML"
      :class="[proxyClasses.content, proxyClasses.contentHtml]"
      v-html="proxyText" />
    <div
      v-else
      :class="[proxyClasses.content, proxyClasses.contentText]">
      {{ proxyText }}
    </div>
    <slot>
      <button
        v-if="showButton"
        :class="[proxyClasses.button, proxyButtonClass]"
        @click.prevent="toggle">
        {{ buttonTitle }}
      </button>
    </slot>
  </div>
</template>

<script lang="ts">
import { computed, defineComponent, PropType } from 'vue';
import htmlTruncate from 'html-truncate';
import sanitizeHtml, { IOptions } from 'sanitize-html';
import { Classes, Buttons, Type } from './types';
import { defaultClasses, defaultButtons, HTML } from './const';

export default defineComponent({
  name: 'VueTruncateHtml',
  props: {
    modelValue: {
      type: Boolean,
      required: true,
    },
    text: {
      type: String,
      default: '',
    },
    length: {
      type: Number,
      default: 100,
    },
    hideButton: {
      type: Boolean,
      default: false,
    },
    type: {
      type: String as PropType<Type>,
      default: 'text',
    },
    buttons: {
      type: Object as PropType<Buttons>,
      default: () => defaultButtons,
    },
    classes: {
      type: Object as PropType<Classes>,
      default: () => defaultClasses,
    },
    sanitizeOptions: {
      type: Object as PropType<IOptions>,
      default: undefined,
    },
  },
  emits: ['update:modelValue'],
  setup(props, { emit }) {
    const isTruncated = computed({
      get: () => props.modelValue,
      set: (value) => emit('update:modelValue', value),
    });

    const isHTML = computed(() => props.type === HTML);

    const textLength = computed(() => {
      const text = isHTML.value ? props.text.replace(/<[^>]*>/g, '') : props.text;

      return text.length;
    });

    const showButton = computed(() => !props.hideButton && textLength.value > props.length);

    const sanitizedHtmlOrText = computed(() => (isHTML.value
      ? sanitizeHtml(props.text, props.sanitizeOptions)
      : props.text));
    const truncatedHtmlOrText = computed(() => (
      isHTML.value
        ? htmlTruncate(sanitizedHtmlOrText.value, props.length)
        : sanitizedHtmlOrText.value.substring(0, props.length)
    ));

    const proxyText = computed(() => (isTruncated.value ? truncatedHtmlOrText.value : sanitizedHtmlOrText.value));

    const buttonTitle = computed(() => (
      isTruncated.value
        ? props.buttons.more ?? defaultButtons.more
        : props.buttons.less ?? defaultButtons.less
    ));

    const proxyClasses = computed(() => ({
      container: props.classes?.container ?? defaultClasses.container,
      content: props.classes?.content ?? defaultClasses.content,
      contentHtml: props.classes?.contentHtml ?? defaultClasses.contentHtml,
      contentText: props.classes?.contentText ?? defaultClasses.contentText,
      button: props.classes?.button ?? defaultClasses.button,
      buttonMore: props.classes?.buttonMore ?? defaultClasses.buttonMore,
      buttonLess: props.classes?.buttonLess ?? defaultClasses.buttonLess,
    }));

    const proxyButtonClass = computed(() => (isTruncated.value ? proxyClasses.value.buttonMore : proxyClasses.value.buttonLess));

    const toggle = () => {
      isTruncated.value = !isTruncated.value;
    };

    return {
      isHTML,
      showButton,
      proxyButtonClass,
      proxyText,
      buttonTitle,
      proxyClasses,
      toggle,
    };
  },
});
</script>
