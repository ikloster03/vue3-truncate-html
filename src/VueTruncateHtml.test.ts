import { mount } from '@vue/test-utils';
import { axe, toHaveNoViolations } from 'jest-axe';
import VueTruncateHtml from './VueTruncateHtml.vue';

expect.extend(toHaveNoViolations);

const TEXT = `
      Lorem ipsum dolor sit amet, consectetur adipisicing elit.
      Adipisci animi autem beatae consequuntur debitis delectus deleniti ducimus enim,
      facere hic id impedit labore laboriosam magni molestiae nemo non numquam officiis porro,
      quibusdam tempora totam vel voluptate voluptatem voluptatum. Ad adipisci architecto,
      beatae blanditiis corporis cumque dolor, eaque excepturi exercitationem magnam nihil optio perferendis perspiciatis qui quis,
      `;

const HTML = `
      <p>
        <b>Lorem ipsum dolor sit amet, consectetur adipisicing elit.</b>
        <ul>
          <li>
              <a href="https://google.com">Google.com</a>
              Adipisci animi autem beatae consequuntur debitis delectus deleniti ducimus enim,
          </li>
          <li>facere hic id impedit labore laboriosam magni molestiae nemo non numquam officiis porro,</li>
          <li>quibusdam tempora totam vel voluptate voluptatem voluptatum. Ad adipisci architecto,</li>
        </ul>
        <i>beatae blanditiis corporis cumque dolor</i>, eaque excepturi exercitationem magnam nihil optio perferendis perspiciatis qui quis,
      </p>
     `;

const textComponentCase = {
  data() {
    return { isTruncated: true, text: TEXT };
  },
  template: '<vue-truncate-html v-model="isTruncated" :text="text"></vue-truncate-html>',
  components: { 'vue-truncate-html': VueTruncateHtml },
};

const htmlComponentCase = {
  data() {
    return { isTruncated: true, text: HTML };
  },
  template: '<vue-truncate-html v-model="isTruncated" :text="text" type="html"></vue-truncate-html>',
  components: { 'vue-truncate-html': VueTruncateHtml },
};

test('mount without props', async () => {
  const wrapper = mount(VueTruncateHtml);

  // console.log(wrapper.html());
  expect(wrapper.find('.vue-truncate-html').exists()).toBe(true);
  expect(wrapper.find('.vue-truncate-html__content').exists()).toBe(true);
  expect(wrapper.find('.vue-truncate-html__button').exists()).toBe(false);
});

it('renders correctly', () => {
  const wrapper = mount(textComponentCase);

  expect(wrapper.element).toMatchSnapshot();
});

it('a11y is ok', async () => {
  const wrapper = mount(textComponentCase);

  expect(await axe(wrapper.element, {
    rules: {
      region: { enabled: false },
    },
  })).toHaveNoViolations();
});

test('mount with text', async () => {
  const wrapper = mount(textComponentCase);

  expect(wrapper.find('.vue-truncate-html__content_text').exists()).toBe(true);
  expect(wrapper.find('.vue-truncate-html__content_html').exists()).toBe(false);
});

test('mount with html', async () => {
  const wrapper = mount(htmlComponentCase);

  expect(wrapper.find('.vue-truncate-html__content_html').exists()).toBe(true);
  expect(wrapper.find('.vue-truncate-html__content_text').exists()).toBe(false);
});

test('text is empty', async () => {
  const wrapper = mount({
    ...textComponentCase,
    template: '<vue-truncate-html v-model="isTruncated"></vue-truncate-html>',
  });

  expect(wrapper.find('.vue-truncate-html__content').text()).toBe('');
  expect(wrapper.find('.vue-truncate-html__button').exists()).toBe(false);
});

test('button: more/less', async () => {
  const wrapper = mount(textComponentCase);

  expect(wrapper.find('.vue-truncate-html__button').exists()).toBe(true);

  expect(wrapper.find('.vue-truncate-html__button_more').exists()).toBe(true);
  expect(wrapper.find('.vue-truncate-html__button_less').exists()).toBe(false);

  await wrapper.find('.vue-truncate-html__button').trigger('click');

  expect(wrapper.find('.vue-truncate-html__button_more').exists()).toBe(false);
  expect(wrapper.find('.vue-truncate-html__button_less').exists()).toBe(true);
});

test('button: force hide', async () => {
  const wrapper = mount({
    ...textComponentCase,
    template: '<vue-truncate-html v-model="isTruncated" :text="text" :hide-button="true"></vue-truncate-html>',
  });

  expect(wrapper.find('.vue-truncate-html__button').exists()).toBe(false);
});

test('text length: show button', async () => {
  const wrapper = mount({
    ...textComponentCase,
    template: '<vue-truncate-html v-model="isTruncated" :text="text" :length="text.length - 1"></vue-truncate-html>',
  });

  expect(wrapper.find('.vue-truncate-html__button').exists()).toBe(true);
  expect(wrapper.find('.vue-truncate-html__button_more').exists()).toBe(true);
  expect(wrapper.find('.vue-truncate-html__button_less').exists()).toBe(false);
});

test('text length: hide button', async () => {
  const wrapper = mount({
    ...textComponentCase,
    template: '<vue-truncate-html v-model="isTruncated"></vue-truncate-html>',
  });

  expect(wrapper.find('.vue-truncate-html__button').exists()).toBe(false);
  expect(wrapper.find('.vue-truncate-html__button_more').exists()).toBe(false);
  expect(wrapper.find('.vue-truncate-html__button_less').exists()).toBe(false);
});

test('type html, but real text is just text', async () => {
  const wrapper = mount({
    ...textComponentCase,
    template: '<vue-truncate-html v-model="isTruncated" :text="text" type="html"></vue-truncate-html>',
  });

  expect(wrapper.find('.vue-truncate-html__content_html').exists()).toBe(true);
});

test('type text, real text is html', async () => {
  const wrapper = mount({
    ...htmlComponentCase,
    template: '<vue-truncate-html v-model="isTruncated" :text="text" type="text"></vue-truncate-html>',
  });

  expect(wrapper.find('.vue-truncate-html__content_text').exists()).toBe(true);
});
//
// test('change more/less name', async () => {
//   expect(false).toBe(true);
// });
//
// test('custom button more/less with slot', async () => {
//   expect(false).toBe(true);
// });
//
// test('custom classes', async () => {
//   expect(false).toBe(true);
// });
//
// test('custom sanitize options', async () => {
//   expect(false).toBe(true);
// });
