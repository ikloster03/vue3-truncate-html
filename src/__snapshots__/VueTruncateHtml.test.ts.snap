// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<div
  class="vue-truncate-html"
>
  <div
    class="vue-truncate-html__content vue-truncate-html__content_text"
  >
    
      Lorem ipsum dolor sit amet, consectetur adipisicing elit.
      Adipisci animi autem beatae c
  </div>
  
  <button
    class="vue-truncate-html__button vue-truncate-html__button_more"
  >
    Read More
  </button>
  
</div>
`;
